models/vrp.rb

```diff
@@ -337,49 +337,96 @@ def self.remove_unecessary_units(hash)
+      return unless element[:unavailable_date_ranges]
 
-        s[:unavailable_visit_day_indices] = [] unless s[:unavailable_visit_day_indices]
-        s[:unavailable_visit_day_indices] += s[:unavailable_visit_day_date].to_a.collect{ |unavailable_date|
-          (unavailable_date.to_date - hash[:configuration][:schedule][:range_date][:start]).to_i + start_indice
-        }.compact
-        s.delete(:unavailable_visit_day_date)
+      element[:unavailable_index_ranges] ||= []
+      element[:unavailable_index_ranges] += element[:unavailable_date_ranges].collect{ |range|
+        {
+          start: (range[:start].to_date - hash[:configuration][:schedule][:range_date][:start]).to_i + start_index,
+          end: (range[:end].to_date - hash[:configuration][:schedule][:range_date][:start]).to_i + start_index,
+        }
       }
+      element.delete(:unavailable_date_ranges)
+    end
 
-      # remove vehicle unavailable_work_date
-      hash[:vehicles].each{ |vehicle|
-        next if vehicle[:unavailable_work_date].to_a.empty?
+    def self.collect_unvaible_day_indices(element, type, start_index)
+      return [] unless element[:unavailable_index_ranges].to_a.size.positive?
+
+      unavailable_indices =
+        case type
+        when :visit
+          :unavailable_visit_day_indices
+        when :vehicle
+          :unavailable_work_day_indices
+        when :schedule
+          :unavailable_indices
+        end
 
-        vehicle[:unavailable_work_day_indices] = [] unless vehicle[:unavailable_work_day_indices]
-        vehicle[:unavailable_work_day_indices] += vehicle[:unavailable_work_date].collect{ |unavailable_date|
-          (unavailable_date.to_date - hash[:configuration][:schedule][:range_date][:start]).to_i + start_indice
-        }.compact
-        vehicle.delete(:unavailable_work_date)
-      }
+      element[unavailable_indices] += element[:unavailable_index_ranges].collect{ |range|
+        (range[:start]..range[:end]).to_a
+      }.flatten.uniq.select{ |day| day >= start_index }
+      element.delete(:unavailable_index_ranges)
+    end
+
+    def self.deduce_unavailable_days(hash, element, start_index, type)
+      convert_availability_dates_into_indices(element, hash, start_index, type)
+      collect_unvaible_day_indices(element, type, start_index)
+    end
 
-      # remove schedule unavailable_date
+    def self.generate_schedule_indices_from_date(hash)
+      return hash if !hash[:configuration] || !hash[:configuration][:schedule]
+
+      if hash[:configuration][:schedule][:range_indices] &&
+         hash[:services]&.any?{ |s| s[:unavailable_date_ranges] }
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          'Date intervals are not compatible with schedule range indices'
+        )
+      end
+
+      start_index, end_index =
+        if hash[:configuration][:schedule][:range_indices]
+          hash[:configuration][:schedule][:range_indices][:start]
```

Comment:
in this case, isn't `end_index` set to nil ?

Reply:
bien vu ! if this didn't raise an error during testing this means we have no test passing through this function, @fonsecadeline this might be a good occasion to add a unit test on `generate_schedule_indices_from_date(hash)` or `filter(hash)`

Reply:
yes it is, but that is expected because it won't be used in this case.

Reply:
we do test this case

Reply:
should I return nil explicitly ?

Reply:
But, you put `end_index` in `hash[:configuration][:schedule][:range_indices]`

Isn't this value used ?

Reply:
! But you do `return hash if hash[:configuration][:schedule][:range_indices]` if you have a range indices

Reply:
Right, so when we have range_indices we do not need end_index. 
Anyway, I am changing this to ensure unavailable days are between schedule start and schedule end so I will return end_index in this case too.